namcap.1
.TH namcap 1 "July 24, 2007" "namcap 2.0" "User Commands"
.SH NAME
namcap \- package analysis utility
.SH SYNOPSIS
\fBnamcap [options] <package|PKGBUILD> [package|PKGBUILD] ...
.SH DESCRIPTION
.PP
\fBnamcap\fP is a \fIpackage analysis\fP utility that looks for problems with archlinux packages or their PKGBUILD files.  It can apply rules to the file list, the files themselves, or individual PKGBUILD files.
.PP
Rules return lists of messages.  Each message can be one of three types: error, warning, or information (think of them as notes or comments).  Errors (designated by 'E:') are things that namcap is very sure are wrong and need to be fixed.  Warnings (designated by 'W:') are things that namcap thinks should be changed but if you know what you're doing then you can leave them.  Information (designated 'I:') are only shown when you use the info arguement.  Information messages give information that might be helpful but isn't anything that needs changing.
.SH OPTIONS
.TP
.B "\-i, \-\-info"
display information messages
.TP
\fB\-r\fR RULELIST, \fB\-\-rules=\fRRULELIST
only apply RULELIST rules to the package
.IP
RULELIST is a comma-seperated list of rule names; if RULELIST=list then namcap returns a list of valid rules and their descriptions
.SH RULES
.TP
.B arrays
arrays verifies that array variables are actually arrays and not strings
.TP
.B badbackups
badbackups verifies that backup entries don't have a / at the beginning of them
.TP
.B carch
carch looks for references to i686 and i586 and tells you to use $CARCH instead
.TP
.B capsnames
capsnames checks a PKGBUILD to verify the package name does not include upper case characters
.TP
.B capsnamespkg
capsnamespkg checks a package to verify the package name does not include upper case characters
.TP
.B depends
depends runs ldd on all executables, gets the link-level dependencies, finds the smallest subset of dependencies that cover the link-level dependencies, and compares that list to the depends of the package.  It returns messages in three cases: dependency detected and not included, dependency included but already satisfied, and dependency included and not needed.  These suggestions are just guidelines and all package builders should take this into account (ie. you're smarter than namcap is)

Some cases where namcap fails are dlopen() and obscure links.  dlopen()'d libraries don't show up because they are loaded at run time: in the case of a prgram that loads plugins.  Obscure links are the cases where only a small portion of the package needs something to run; usually, the small portion won't be accesed unless that thing is installed (i.e. a java plugin for mozilla included in the java package).
.TP
.B directoryname
directoryname checks all of the directory names for standard directories (ie. /etc, /usr/bin, /usr/lib).  All files in directories that are non-standard are returned
.TP
.B emptydir
emptydir looks for directories in a package that don't contain any files
.TP
.B fhsinfopages
fhsinfopages verifies correct installation of info pages
.TP
.B fhsmanpages
fhsmanpages verifies correct installation of man pages
.TP
.B fileownership
fileownership verifies that all files are owned by root:root and not some other user or group
.TP
.B gnomemenu
gnomemenu looks for file to be stored in /usr/share/gnome/apps
.TP
.B gnomemime
gnomemime checks for generated GNOME mime files
.TP
.B infodirectory
infodirectory verifies that a package does not contain the info directory file (/usr/share/info/dir)
.TP
.B invalidstartdir
Any PKGBUILDs that refer to files not stored in $startdir/pkg or $startdir/src are incorrect
.TP
.B libtool
libtool checks for libtool (*.la) files
.TP
.B license
license verifies that the licenses variable has been filled in in a PKGBUILD
.TP
.B licensepkg
licensepkg verifies that the licenses variable has been filled in in a package
.TP
.B md5sums
md5sums makes sure that a PKGBUILD includes the md5sums
.TP
.B perllocal
perllocal searches for perllocal.pod.  perllocal.pod is a nasty file that's included during most perl module installations
.TP
.B permissions
permissions checks basic file and and directory permissions.  It returns warnings about world readable bits not being set, world writable bits being set, and world executable bit not being set for directories
.TP
.B scrollkeeper
scrollkeeper verifies that there aren't any scrollkeeper directories
.TP
.B sfurl
sfurl verifies that a PKGBUILD is downloading from a specific sourceforge mirror
.TP
.B symlink
symlink points out symlinks that point to $startdir/pkg instead of where the files will actually be installed to
.TP
.B tags
tags is for package maintainers.  It verifies that we have remembered to add a Maintainer and Contributor comment.
.TP
.B url
url verifies that we have a url variable set in the PKGBUILD
.TP
.B urlpkg
url verifies that we have a url variable set in the package file
.SH EXAMPLES
.TP
.B namcap foo.pkg.tar.gz
apply all rules to foo.pkg.tar.gz
.TP
.B namcap -r depends,symlink foo.pkg.tar.gz
apply the depends and symlink rules to foo.pkg.tar.gz
.TP
.B namcap PKGBUILD
apply all PKGBUILD based rules to the file PKGBUILD
.TP
.B namcap -r list
list all of the available rules
.SH COPYRIGHT
Copyright \(co 2003-2007 Jason Chu
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH AUTHORS
.nf
Jason Chu <jason@archlinux.org>
JJDaNiMoTh <jjdanimoth@gmail.com>
Jesse Young <jesseyoung@gmail.com>
.fi


namcap.py
#! /usr/bin/env python
# 
# namcap - A Pacman package analyzer
# Copyright (C) 2003-2007 Jason Chu <jason@archlinux.org>
# 
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation; either version 2 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software
#   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
# 

# System wide global stuff
import sys, os, os.path, imp, getopt, types, tarfile, re, string, Namcap, pacman
import shutil

sandbox_directory = '/tmp/namcap.' + str(os.getpid())

# Functions
# Return all possible modules (rules)
def get_modules():
	return Namcap.__all__

# Display usage information
def usage():
	print "usage: " + sys.argv[0] + " [-r rulelist | --rules=rulelist] [-i | --info] package .."
	print "       -r list    : returns list of available rules"
	print "       -i         : prints information responses from rules"
	sys.exit(2)

# Is the package a valid package file?
def verify_package(filename):
	if not os.path.isfile(filename):
		return 0
	if not tarfile.is_tarfile(filename):
		return 0
	try:
		tar = tarfile.open(package, "r")
		if not tar:
			return 0
		if not '.PKGINFO' in tar.getnames():
			tar.close()
			return 0
	except IOError:
		if tar:
			tar.close()
		return 0
	return tar

def process_tags(filename="/usr/share/namcap/tags"):
	tags = {}
	f = open(filename)
	for i in f.readlines():
		if i[0] == "#" or i.strip() == "": continue
		tagdata = i[:-1].split("::")
		tags[tagdata[0].strip()] = tagdata[1].strip()

	return tags

# Main
modules = get_modules()
tags = process_tags()
info_reporting = 0

# get our options and process them
try:
	optlist, args = getopt.getopt(sys.argv[1:], "ihmr:", ["rules=","info","help","machine-readable"])
except getopt.GetoptError:
	usage()

active_modules = []
m = lambda s: tags[s]

for i, k in optlist:
	if i in ('-r', '--rules') and active_modules == []:
		if k == 'list':
			print "-"*20 + " Namcap rule list " + "-"*20
			for j in modules:
				print string.ljust(j, 20) + ": " + __import__('Namcap.' + j, globals(), locals(), [Namcap]).package().long_name()
			sys.exit(2)
			
		module_list = k.split(',')
		for j in module_list:
			if j in modules:
				active_modules.append(j)
			else:
				print "Error: Rule '" + j + "' does not exist"
				usage()
	if i in ('-i', '--info'):
		info_reporting = 1
	if i in ('-m', '--machine-readable'):
		machine_readable = 1
		m = lambda s: s
	if i in ('-h', '--help'):
		usage()

# If there are no args, print usage
if (args == []):
	usage()

packages = args

# Go through each package, get the info, and apply the rules
for package in packages:
	extracted = 0
	if not os.access(package, os.R_OK):
		print "Error: Problem reading " + package
		usage()

	if package[-7:] == '.tar.gz':
		pkgtar = verify_package(package)

		if not pkgtar:
			print "Error: " + package + " is not a package"
			if len(packages) > 1:
				continue

		pkginfo = pacman.load(package)

		# No rules selected?  Then select them all!
		if active_modules == []:
			active_modules = modules

		# Loop through each one, load them apply if possible
		for i in active_modules:
			cur_class = __import__('Namcap.' + i, globals(), locals(), [Namcap])
			pkg = cur_class.package()
			ret = [[],[],[]]
			if pkg.type() == "tarball":
				if pkg.prereq() == "extract":
					# If it's not extracted, then extract it and then analyze the package
					if not extracted:
						os.mkdir(sandbox_directory)
						for j in pkgtar.getmembers():
							pkgtar.extract(j, sandbox_directory)
						extracted = 1
					ret = pkg.analyze(pkginfo, sandbox_directory)
				elif pkg.prereq() == "pkg":
					ret = pkg.analyze(pkginfo, None)
				elif pkg.prereq() == "tar":
					ret = pkg.analyze(pkginfo, pkgtar)
				else:
					ret = [['Error running rule (' + i + ')'],[],[]]

				# Output the three types of messages
				if ret[0] != []:
					for j in ret[0]:
						print string.ljust(pkginfo.name, 10) + " E: " + m(j[0]) % j[1]
				if ret[1] != []:
					for j in ret[1]:
						print string.ljust(pkginfo.name, 10) + " W: " +  m(j[0]) % j[1]
				if ret[2] != [] and info_reporting:
					for j in ret[2]:
						print string.ljust(pkginfo.name, 10) + " I: " +  m(j[0]) % j[1]


		# Clean up if we extracted anything
		if extracted:
			shutil.rmtree(sandbox_directory)
	elif package[-8:] == 'PKGBUILD':
		# We might want to do some verifying in here... but really... isn't that what pacman.load is for?
		pkginfo = pacman.load(package)

		if pkginfo == None:
			print "Error: " + package + " is not a valid PKGBUILD"
			continue

		if active_modules == []:
			active_modules = modules

		for i in active_modules:
			cur_class = __import__('Namcap.' + i, globals(), locals(), [Namcap])
			pkg = cur_class.package()
			ret = [[],[],[]]
			if pkg.type() == "pkgbuild":
				ret = pkg.analyze(pkginfo, package)

			# Output the PKGBUILD messages
			if ret[0] != []:
				for j in ret[0]:
					print string.ljust("PKGBUILD (" + pkginfo.name + ")", 20) + " E: " + m(j[0]) % j[1]
			if ret[1] != []:
				for j in ret[1]:
					print string.ljust("PKGBUILD (" + pkginfo.name + ")", 20) + " W: " + m(j[0]) % j[1]
			if ret[2] != [] and info_reporting:
				for j in ret[2]:
					print string.ljust("PKGBUILD (" + pkginfo.name + ")", 20) + " I: " + m(j[0]) % j[1]
# vim: set ts=4 sw=4 noet:
